[Jason Baek] iP

bkjwjason opened the pull request:
## BaekBot

> "Use BaekBot and your productivity will increase by 500%" - [Anonymous](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

This is the latest release of ~Duke~ **BaekBot**, the <sup>superior</sup> Task Managing Bot! You will learn to develop

- Task Logging Skills
- Responsibility for your work
- And lastly, a sense of accomplishment whenever you finish a task!

### BaekBot Functions

BaekBot supports the following functions.

1. Adding a Task of different types (Event, Deadline, Todo)
2. Marking/Unmarking a Task as done
3. Removing a task from the list
4. Saving your tasklist

What do I mean by Marking/Unmarking a Task as done? It is simply the below expressed in the CLI.

- [x] Task 1
- [x] Task 2
- [ ] Task 3

Use relevant commands and see your tasks being marked/unmarked just like the above! 👍 

For Java Specifics, the bot runs from the main method below

```java
    public static void main(String[] args){
        Duke duke = new Duke();
        duke.start();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

bkjwjason committed:
add intro message

bkjwjason committed:
add echo function

bkjwjason committed:
add task listing function

bkjwjason committed:
add task marking function

bkjwjason committed:
add task categories

bkjwjason committed:
add testcase

bkjwjason committed:
add exception handling

bkjwjason committed:
add deletion functionality

bkjwjason committed:
read/write to local storage

bkjwjason committed:
Merge branch 'branch-Level-7'
Merge Level-7 branch for read/write to local storage

bkjwjason committed:
add storage of dates as DateTime objects

bkjwjason committed:
Merge branch 'branch-Level-8'
Merge branch-Level-8 for storage of DateTime objects

bkjwjason committed:
abstract functions and variables into more classes

bkjwjason committed:
Merge branch 'A-MoreOOP'
Merge A-MoreOOP branch for greater abstraction of classes

bkjwjason committed:
divide classes into relevant packages

bkjwjason committed:
Merge branch 'A-Packages'
merge with A-Packages branch to refactor classes

bkjwjason committed:
Merge tag 'A-Packages'
merge branch A-Packages to refactor classes into relevant packages.

bkjwjason committed:
Merge branch 'A-Gradle'
Merge Gradle Support into main

bkjwjason committed:
add JUnit Tests for Parser, Todo and Deadline

bkjwjason committed:
Merge branch 'A-JUnit'
Merge with branch A-JUnit to integrate Unit Testing

bkjwjason committed:
Add header comments to all classes and methods

bkjwjason committed:
Tweak code layout to meet coding standards

bkjwjason committed:
Merge branch 'A-JavaDoc'
Merge with branch A-JavaDoc for JavaDocs on all classes and methods

bkjwjason committed:
Merge branch 'A-CodingStandard'
Merge with branch A-CodingStandard to make code align with coding
standards.

bkjwjason committed:
Add ability to find Tasks based on keyword

bkjwjason committed:
Merge branch 'Level-9'
Merge with branch Level-9 to add searching function based on keyword

frrrrry commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,96 @@
+ * and read/write to the data file when the bot it used.
+ */
+public class Storage {
+    private static final String FOLDER_PATH = "./data";
```
I like how the constant variables are all in uppercase with a underscore separating, following the java coding standards. :+1:

frrrrry commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,157 @@
+     * @param task Task that was added to the list.
+     * @param count Total count of tasks in the list after addition.
+     */
+    public void addTaskMessage(Task task, int count) {
```
I like how you name your methods as verbs as they are very easy and clear to understand. :+1:

frrrrry commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,15 @@
+/**
+ * DukeException is a custom Exception made for the Duke Application
+ */
+public class DukeException extends Exception{
```
Should there be more exception classes to specify more specific errors instead of just one general exception class?

bkjwjason replied:
Great suggestion, thank you!

frrrrry commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,106 @@
+        String mainCommand = spacedCommand[0];
+        Task task;
+        try {
+            switch (mainCommand) {
```
I like how the indentations of the switch case statements follow the java coding standards as its a pretty common mistake to forget that the switch case statements such indentation style.

frrrrry commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,129 @@
+ * TaskList is a class containing Tasks. Through the TaskList, users are able to
+ * add, delete, view as well as mark a certain task as done.
+ */
+public class TaskList {
```
Should this class be placed in the task package or another package?

frrrrry commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    /**
+     * Modifies a task at given index by either marking it as done or not done.
```
Javadocs comments are clear and easy to understand 👍

frrrrry reviewed (commented):
LGTM. I like how you adhered to java coding standards with regards to method naming and indentation of switch case statements. Javadocs comments are easy to understand too. An improvement can be to add more specific exception classes for error handling. :stuck_out_tongue_closed_eyes:

bkjwjason committed:
Add relevant JavaFX controllers for GUI

bkjwjason committed:
Change method signature of Parse method

bkjwjason committed:
Use FXML for scene building

bkjwjason committed:
Add initial greet message

bkjwjason committed:
Add javadocs for new classes

bkjwjason committed:
Fix styling issues

bkjwjason committed:
Merge branch 'Level-10'
Merge with branch Level-10 which allows for GUI based interaction.

bkjwjason committed:
Display error messages as strings

peasantbird commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    /**
+     * Sets input to be the variable duke.
+     * @param d to be set to the variable duke.
```
Perhaps there should be a space between these two lines in the javadoc?

peasantbird commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,119 @@
+        Task task;
+        String output = null;
+        try {
+            switch (mainCommand) {
```
I like the indentations of this switch-case block that do not make it look like an arrowhead!

peasantbird commented on src/test/java/duke/task/TodoTest.java:
```diff
@@ -0,0 +1,23 @@
+public class TodoTest {
+
+    @Test
+    public void toStringTest() {
```
Perhaps this should be testToString()?

bkjwjason replied:
I agree, that is a better naming convention

peasantbird commented on src/test/java/duke/task/TodoTest.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    @Test
+    public void markDoneTest() {
```
Perhaps this should be testMarkDone() so that the naming for tests is more standardized?

bkjwjason replied:
Yes i agree with your suggestion, thank you!

peasantbird reviewed (commented):
LGTM! Just a few nitpicking on the naming of JUnit tests and line breaks, but overall very well-written and clean!

bkjwjason committed:
Fix Junit test not catching errors

bkjwjason committed:
Add assert statements to certain functions

Assert statements help us document important
assumptions that should hold in certain
points of code. Assert statements have been added
to modification of tasklist functions such as
addTask and findTask as retrieval of task
from TaskList is assumed to not be nulltype. This
will help allow us to hold our assumptions true.

bkjwjason committed:
Merge pull request #1 from bkjwjason/A-Assertions

Add assert statements to certain functions

bkjwjason committed:
Change name of test functions

Change name of test functions in TodoTest in
test folder to adhere to naming conventions.
Naming conventions help coders to focus more
on important issues than syntax and naming
standards. It is important to keep naming
standardised, hence it was changed.

bkjwjason committed:
Merge pull request #2 from bkjwjason/A-CodeQuality

Change name of test functions

bkjwjason committed:
Add function to check for duplicate Tasks

bkjwjason committed:
Merge pull request #3 from bkjwjason/C-DetectDuplicates

Add function to check for duplicate Tasks

bkjwjason committed:
Refactor Parse function to adhere to SLAP

bkjwjason committed:
Add javadocs for refactored parse functions

bkjwjason committed:
Modify javadoc to adhere to coding standards

bkjwjason committed:
Add UserGuide

bkjwjason committed:
Fix UserGuide typo
